File: src/csharp/oop/ParkingLot/Parking/Fees/FeeCalculator.cs
namespace ParkingLot.Parking.Fees;

public sealed class FeeCalculator(IReadOnlyCollection<IHourRateStrategy> strategies) : IFeeCalculator
{
    public decimal Calculate(Ticket ticket)
    {
        if (ticket.Invalid)
        {
            return 0;
        }
        
        decimal rate = 0;
        foreach (var strategy in strategies)
        {
            rate = strategy.Calculate(ticket, ticket.ParkTime, rate);
        }

        TimeSpan parkDuration = ticket.LeaveTime!.Value - ticket.ParkTime;
        var hours = parkDuration.Hours > 0 ? parkDuration.Hours : 1;
        
        return hours * rate;
    }
}

File: src/python/oop/ParkingLot/parking/fee/fee_calculator.py
from .rate_strategy_base import RateStrategyBase
from ..ticket import Ticket

class FeeCalculator:
    def __init__(self, strategies: list[RateStrategyBase]):
        self.__strategies = strategies

    def calculate(self, ticket: Ticket) -> float:
        rate = 0.0
        for strategy in self.__strategies:
            rate = strategy.calculate(ticket, rate)

        duration = ticket.leave_time - ticket.park_time
        hours = duration.total_seconds() / 3600

        hours = hours if hours >= 1 else 1

        return hours * rate
